File: php/nodes/save_m3u_file.php
<?php

function save_m3u_file(array $edges, array $utils = []): callable {
    $utils += ['write_text_file' => fn($p, $c) => file_put_contents($p, $c)];

    return function ($state) use ($edges, $utils): Next {
        if (empty($state['m3u_file_path'])) {
            throw new \RuntimeException('M3U file path is not defined.');
        }

        $content = implode("\n", $state['urls']);
        $utils['write_text_file']($state['m3u_file_path'], $content);
        echo "Saved resolved URLs to {$state['m3u_file_path']}\n";
        return new Next($edges['on_success'], $state);
    };
}

File: ruby/nodes/save_m3u_file.rb
# frozen_string_literal: true

require_relative '../lib/ambler'

module SaveM3uFile
  DEFAULT_UTILS = { write_file: ->(path, content) { File.write(path, content) } }.freeze

  def self.node(edges, utils = DEFAULT_UTILS)
    ->(state) {
      raise 'M3U file path is not defined.' if state[:m3u_file_path].nil?

      content = state[:urls].join("\n")
      utils[:write_file].call(state[:m3u_file_path], content)
      puts "Saved resolved URLs to #{state[:m3u_file_path]}"
      Ambler::Next.new(edges[:on_success], state)
    }
  end
end
